li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/metadataservice/MarioMetadataServiceClient.java

```diff
@@ -0,0 +1,159 @@
+    _clientIdToClusterGroupMap = new HashMap<>();
+  }
+
+  @Override
+  public void configure(Map<String, ?> configs) {
+  }
+
+  @Override
+  public UUID registerFederatedClient(ClusterGroupDescriptor clusterGroup, Map<String, ?> configs, int timeoutMs) {
+    if (clusterGroup == null) {
+      throw new IllegalArgumentException("cluster group cannot be null");
+    }
+
+    // This is a temporary hack until Mario supports client registration.
+    UUID clientId = UUID.randomUUID();
+    _clientIdToClusterGroupMap.put(clientId, clusterGroup);
+    return clientId;
+  }
+
+  @Override
+  public ClusterDescriptor getClusterForTopic(UUID clientId, String topicName, int timeoutMs)
+      throws MetadataServiceClientException {
+    Set<ClusterDescriptor> clusters = getClusterMapForTopics(clientId, new HashSet<>(Arrays.asList(topicName)),
+        timeoutMs).get(topicName);
+    if (clusters == null || clusters.isEmpty()) {
+      return null;
+    }
+    // Unless topic move is in progress, a topic must exist in only one cluster.
+    if (clusters.size() > 1) {
+      throw new IllegalStateException("topic " + topicName + " exists in more than one cluster " + clusters +
+          " in cluster group " + _clientIdToClusterGroupMap.get(clientId).name());
+    }
+    return clusters.iterator().next();
+  }
+
+  @Override
+  public Map<TopicPartition, ClusterDescriptor> getClustersForTopicPartitions(UUID clientId,
+      Collection<TopicPartition> topicPartitions, int timeoutMs) throws MetadataServiceClientException {
+    // Create a set of topic names out of topicPartitions.
+    Set<String> topicNames = new HashSet<>();
+    for (Iterator<TopicPartition> it = topicPartitions.iterator(); it.hasNext(); ) {
+      topicNames.add(it.next().topic());
+    }
+
+    Map<String, Set<ClusterDescriptor>> topicToClusterMap = getClusterMapForTopics(clientId, topicNames, timeoutMs);
+    Map<TopicPartition, ClusterDescriptor> topicPartitionToClusterMap = new HashMap<>();
+    for (Iterator<TopicPartition> it = topicPartitions.iterator(); it.hasNext(); ) {
+      TopicPartition topicPartition = it.next();
+      Set<ClusterDescriptor> cluster = topicToClusterMap.get(topicPartition.topic());
+      if (cluster == null || cluster.isEmpty()) {
+        continue;
+      }
+
+      // Unless topic move is in progress, a topic must exist in only one cluster.
+      if (cluster.size() > 1) {
+        throw new IllegalStateException("topic " + topicPartition.topic() + " exists in more than one cluster");
+      }
+
+      topicPartitionToClusterMap.put(topicPartition, cluster.iterator().next());
+    }
+    return topicPartitionToClusterMap;
+  }
+
+  // For given topic names, construct a map from topic name to cluster descriptors where the topic exists. If a topic
+  // does not exist, there will be no entry for that topic in the return map.
+  private Map<String, Set<ClusterDescriptor>> getClusterMapForTopics(UUID clientId, Set<String> topicNames,
```

Comment:
just to clarify - this was meant to allow multiple clients (each with their own ID) to use the same MD service client?

i still think this is a bit of a violation of concerns (why should NDS client keep state per client ID? when does that state expire on client crash?)

Reply:
Based on our discussion, updated the code to pass cluster group to MDSClient to make it stateless.